expression/builtin_time.go

```diff
@@ -2256,160 +2287,420 @@ func isDuration(str string) bool {
+		return "", errors.Trace(err)
 	}
 	arg1time, err := arg1.ConvertToTime(uint8(getFsp(arg1.String())))
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	tmpDuration := arg0.Sub(&arg1time)
+	fsp := types.MaxFsp
+	if tmpDuration.MicroSecond() == 0 {
+		fsp = types.MinFsp
+	}
 	resultDuration, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
 	if err != nil {
-		return result, errors.Trace(err)
-	}
-	if getFsp(d) != 0 {
-		tmpDuration.Fsp = types.MaxFsp
-	} else {
-		tmpDuration.Fsp = types.MinFsp
+		return "", errors.Trace(err)
 	}
-	result.SetString(resultDuration.String())
-	return
+	resultDuration.Fsp = fsp
+	return resultDuration.String(), nil
 }
 
-// strDurationSubDuration subtracts duration from duration string, returns a datum value.
-func strDurationSubDuration(d string, arg1 types.Duration) (result types.Datum, err error) {
-	arg0, err := types.ParseDuration(d, getFsp(d))
+// strDurationSubDuration subtracts duration from duration string, returns a string value.
+func strDurationSubDuration(d string, arg1 types.Duration) (string, error) {
+	arg0, err := types.ParseDuration(d, types.MaxFsp)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	tmpDuration, err := arg0.Sub(arg1)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
-	if getFsp(d) != 0 {
-		tmpDuration.Fsp = types.MaxFsp
-	} else {
+	tmpDuration.Fsp = types.MaxFsp
+	if tmpDuration.MicroSecond() == 0 {
 		tmpDuration.Fsp = types.MinFsp
 	}
-	result.SetString(tmpDuration.String())
-	return
+	return tmpDuration.String(), nil
 }
 
 type addTimeFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *addTimeFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *addTimeFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinAddTimeSig{newBaseBuiltinFunc(args, ctx)}
+	tp1, tp2 := args[0].GetType(), args[1].GetType()
+	var argTp1, argTp2, retTp evalTp
+	switch tp1.Tp {
+	case mysql.TypeDatetime, mysql.TypeTimestamp:
+		argTp1, retTp = tpDatetime, tpDatetime
+	case mysql.TypeDuration:
+		argTp1, retTp = tpDuration, tpDuration
+	case mysql.TypeDate:
+		argTp1, retTp = tpDuration, tpString
+	default:
+		argTp1, retTp = tpString, tpString
+	}
+	switch tp2.Tp {
+	case mysql.TypeDatetime, mysql.TypeDuration:
+		argTp2 = tpDuration
+	default:
+		argTp2 = tpString
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, retTp, argTp1, argTp2)
+	bf.tp.Decimal = tp1.Decimal
+	if retTp == tpString {
+		bf.tp.Tp, bf.tp.Flen, bf.tp.Decimal = mysql.TypeString, mysql.MaxDatetimeWidthWithFsp, types.UnspecifiedLength
+	}
+	switch tp1.Tp {
+	case mysql.TypeDatetime, mysql.TypeTimestamp:
+		switch tp2.Tp {
+		case mysql.TypeDatetime, mysql.TypeDuration:
```

Comment:
if tp2.Tp == mysql.TypeDatetime, we calculate it as tpDuration?
According to the mysql manual, tp2 must be a duration expr.
I'm not sure whether `fieldTp2EvalTp(tp2.Tp) == tpDatetime || fieldTp2EvalTp(tp2.Tp) == tpTimestamp` is true, addtime always return null.

Reply:
The behavior here in Mysql is confusing..
```sql
+-------------------------------------------------------------------------------------------+
| addtime(cast('2017:01:01 01:01:01' as datetime), cast('2017:01:01 01:01:01' as datetime)) |
+-------------------------------------------------------------------------------------------+
| 2017-01-01 02:02:02                                                                       |
+-------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> create table ttt( a datetime);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into ttt values('2017:01:01 01:01:01');
Query OK, 1 row affected (0.01 sec)

mysql> select addtime(a,a) from ttt;
Field   1:  `addtime(a,a)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     19
Max_length: 0
Decimals:   0
Flags:      BINARY 


+--------------+
| addtime(a,a) |
+--------------+
| NULL         |
+--------------+
1 row in set (0.00 sec)
```

Reply:
you may refer to the source code of MySQL.

Reply:
@XuHuaiyu I guess it has something to do with `CAST` in MySQL.  Maybe the optimizer in MySQL does some adaptation work dealing with `CAST` in this condition. I'll take a look in MySQL source code. Some other cases are as follows. 
```sql
mysql> create table dt_test(a datetime, b varchar(100));                                                                                     
Query OK, 0 rows affected (0.01 sec)

mysql> insert into dt_test values('2017:01:01 01:01:01', '2017:01:01 01:01:01');                                                             
Query OK, 1 row affected (0.00 sec)

mysql> select addtime(a, cast(b as datetime)) from dt_test;                                                                                  
Field   1:  `addtime(a, cast(b as datetime))`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     19
Max_length: 19
Decimals:   0
Flags:      BINARY


+---------------------------------+
| addtime(a, cast(b as datetime)) |
+---------------------------------+
| 2017-01-01 02:02:02             |
+---------------------------------+
1 row in set (0.00 sec)

mysql> select subtime(s.a, s.b) from ( select a as a, cast(b as datetime) b from dt_test) s;
Field   1:  `subtime(s.a, s.b)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     19
Max_length: 19
Decimals:   0
Flags:      BINARY 


+---------------------+
| subtime(s.a, s.b)   |
+---------------------+
| 2017-01-01 00:00:00 |
+---------------------+
1 row in set (0.00 sec)

mysql> select subtime(s.a, s.b) from ( select a as a, a as b from dt_test) s;
Field   1:  `subtime(s.a, s.b)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     19
Max_length: 0
Decimals:   0
Flags:      BINARY


+-------------------+
| subtime(s.a, s.b) |
+-------------------+
| NULL              |
+-------------------+
1 row in set (0.00 sec)

```

Reply:
@XuHuaiyu I took a look in MySQL source code.
in `sql/item_timefunc.cc`, see the following funtion:
```c++
bool Item_func_add_time::val_datetime(MYSQL_TIME *time, uint fuzzy_date)
{
  DBUG_ASSERT(fixed == 1);
  MYSQL_TIME l_time1, l_time2;
  bool is_time= 0;
  long days, microseconds;
  longlong seconds;
  int l_sign= sign;

  null_value=0;
  if (cached_field_type == MYSQL_TYPE_DATETIME)  // TIMESTAMP function
  {
    if (get_arg0_date(&l_time1, fuzzy_date) || 
        args[1]->get_time(&l_time2) ||
        l_time1.time_type == MYSQL_TIMESTAMP_TIME || 
        l_time2.time_type != MYSQL_TIMESTAMP_TIME)
      goto null_date;
  }
  else                                // ADDTIME function
  {
    if (args[0]->get_time(&l_time1) || 
        args[1]->get_time(&l_time2) ||
        l_time2.time_type == MYSQL_TIMESTAMP_DATETIME)
      goto null_date;
    is_time= (l_time1.time_type == MYSQL_TIMESTAMP_TIME);
  }
 \\....other codes
```
in sql like 
```sql
select addtime(cast('2017:01:01 01:01:01' as datetime), cast('2017:01:01 01:01:01' as datetime))
```
args[1] is of type `Item_datetime_typecast`, which define can be seen in `sql/item_timefunc.h` as follows:
```c++
class Item_datetime_typecast :public Item_datetime_func
{
\\...
}
```

`Item_datetime_typecast` inherit class `Item_datetime_func`, which implement interface `get_time`:
```c++
class Item_datetime_func :public Item_temporal_func
{
\\....
  bool get_time(MYSQL_TIME *ltime)
  {
    return get_time_from_datetime(ltime);
  }
\\....
}
```
`gettime` is implemented in `sql/item.cc`:
```c++
bool Item::get_time_from_datetime(MYSQL_TIME *ltime)
{
  DBUG_ASSERT(fixed == 1);
  if (get_date(ltime, TIME_FUZZY_DATE))
    return true;
  datetime_to_time(ltime);
  return false;
}
```
and `datetime_to_time` is implemented in `sql/sql_time.h`:
```c++
inline void datetime_to_time(MYSQL_TIME *ltime)
{
  ltime->year= ltime->month= ltime->day= 0;
  ltime->time_type= MYSQL_TIMESTAMP_TIME;
}
```
So, former cases works with a `cast`.

Reply:
@XuHuaiyu I've changed the implementation. If the second parameter is of `DATETIME` type, `NULL` would be returned. so cases with `cast`  like ` addtime(cast('2017:01:01 01:01:01' as datetime), cast('2017:01:01 01:01:01' as datetime))` would return `NULL` too. I think it comply better with the document.